3.12/README.md

```diff
@@ -0,0 +1,144 @@
+# What will 3.12 look like?
+
+## Fundamentals
+
+* We can't speedup opaque blobs of C code.
+    * And we don't want use some elaborate mechanism like PyPy or GraalPython to understand the C code.
+* We can speedup code that runs in the interpreter
+* We can reduce the overhead of memory management, by doing less of it, and by doing it more efficiently.
+* Not doing something at all is always preferable to doing it faster.
+
+### This means that we want:
+
+* C code to consist of builtin functions that are well typed and don't call back in Python (much).
+* Everything else in bytecode (99.9% Python, but we might need a few key trampolines in hand written bytecode).
+* To able to execute Python code faster.
+* To remove incidental overhead due to poor memory layout, reference counting and cycle GC.
+
+## Executing Python faster
+
+This is the most important part of a fast VM, and the most complex (unless we want a fully concurrent, parallel cycle GC).
+Ultimately we will need a JIT compiler, but it is clear from the relatively poor performance of Jython and IronPython that simply compiling to machine code is not magic pixie dust.
+Remembering that not doing something at all is always preferable to doing it faster, we want to optimize by removing overheads, including type-checks, boxing and unboxing, reference counting operations, object allocation.
+
+### Region selection
+
+The simplest region to optimize is a linear trace; all instructions dominate the latter ones, and there are no merge points.
+Traces can be stitched together to form larger regions, or even recompiled as a larger region (see HHVM).
+
+Pyston and Cinder choose single Python functions as their base region of optimization. Even with inlining, this is quite limiting.
+
+My PhD thesis compares a tracing interpreter (HotPy) with a method-at-a-time JIT compiler that did little or no specialization (unladen swallow); the interpreter wins easily.
+
+#### Trace selection
+
+There are two main approaches to tracing in the literature and in industry.
+* Long traces: Traces cover a whole loop, or large number of instructions, as used in PyPy, luajit and tracemonkey.
+* Short traces: Tracelets, or single BB traces, as used in HHVM or YJIT.
+
+Long traces have the advantage that larger regions enable better optimization, and may include a loop which allows loop optimizations.
+Short traces fail to complete much less often, an do not need a custom tracing interpreter.
+
+We are greedy and want the advantages of both!
+We will use the mechanisms for creating short traces, but the type information gathered by the adaptive interpreter will allow us
+to lengthen the traces, without loosing the advantages of short traces.
```

Comment:
How much would tagged pointers help in this regard, both in overhead and in memory usage?